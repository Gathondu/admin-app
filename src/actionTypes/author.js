export const ADD_AUTHOR = "course/ADD_AUTHOR";
export const REMOVE_AUTHOR = "course/REMOVE_AUTHOR";
export const LOAD_AUTHORS_SUCCESS = "course/LOAD_AUTHORS_SUCCESS";
