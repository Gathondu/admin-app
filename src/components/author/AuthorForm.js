import React from 'react';
import PropTypes from 'prop-types';
import TextInput from "../common/TextInput";

const AuthorForm = ({author,errors,onChange,loading, formIsValid, onSave}) => {
    return (
        <form>
            <h1>Manage Author</h1>
            <TextInput
                label="First Name"
                onChange={onChange}
                name="firstName"
                value={author.firstName}
                error={errors.firstName}
                placeholder="Enter First Name"/>
            <TextInput
                label="Last Name"
                onChange={onChange}
                name="lastName"
                value={author.lastName}
                error={errors.lastName}
                placeholder="Enter Course Category"/>
            <input
                type="submit"
                disabled={loading || !formIsValid}
                value={loading ? 'Saving...' : 'Save'}
                className="btn btn-primary"
                onClick={onSave}/>
        </form>
    );
};

AuthorForm.propTypes = {
    author: PropTypes.object.isRequired,
    errors: PropTypes.object.isRequired,
    loading: PropTypes.bool.isRequired,
    formIsValid: PropTypes.bool.isRequired,
    onChange: PropTypes.func.isRequired,
    onSave: PropTypes.func.isRequired
};

export default AuthorForm;
