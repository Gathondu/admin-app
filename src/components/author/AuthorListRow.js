import React from 'react';
import PropTypes from 'prop-types';
import {Link} from "react-router-dom";


const AuthorListRow = ({author, deleteAuthor}) => {
    return (
        <tr>
            <td><i className="fa fa-trash" onClick={() => deleteAuthor(author.id)}/></td>
            <td><Link to={`/author/${author.id}`}>{author.firstName + ' ' + author.lastName}</Link></td>
        </tr>
    );
};

AuthorListRow.propTypes = {
    author: PropTypes.object.isRequired,
    deleteAuthor: PropTypes.func.isRequired
};
export default AuthorListRow;
