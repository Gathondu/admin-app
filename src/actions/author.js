import * as AuthorActionTypes from "../actionTypes/author";
import AuthorApi from "../api/mockAuthor";
import {beginAjaxCall} from './ajaxStatus';

export const addAuthor = author => {
  return { type: AuthorActionTypes.ADD_AUTHOR, author };
};

export const removeAuthor = index => {
  return { type: AuthorActionTypes.REMOVE_AUTHOR, index };
};

export const loadAuthorsSuccess = authors => {
  return { type: AuthorActionTypes.LOAD_AUTHORS_SUCCESS, authors };
};

export function loadAuthors() {
  return function(dispatch) {
    dispatch(beginAjaxCall());
    return AuthorApi.getAllAuthors()
      .then(authors => {
        dispatch(loadAuthorsSuccess(authors));
      })
      .catch(error => {
        throw error;
      });
  };
}
