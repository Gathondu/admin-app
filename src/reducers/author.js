import * as AuthorActionTypes from "../actionTypes/author";
import initialState from "./initialState";

export default function Author(state = initialState.authors, action) {
  switch (action.type) {
    case AuthorActionTypes.ADD_AUTHOR:
      return [...state, action.author];
    case AuthorActionTypes.REMOVE_AUTHOR:
      return [
        ...state.splice(0, action.index),
        ...state.splice(action.index + 1)
      ];
    case AuthorActionTypes.LOAD_AUTHORS_SUCCESS:
      return action.authors;
    default:
      return state;
  }
}
