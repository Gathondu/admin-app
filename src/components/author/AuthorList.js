import React from 'react';
import PropTypes from 'prop-types';
import EmptyPage from "../common/EmptyPage";
import LoadingDots from "../common/LoadingDots";
import AuthorListRow from "./AuthorListRow";

const AuthorList = ({authors, deleteAuthor, loading}) => {
    const authorRows = authors.map((author, index) => (
        <AuthorListRow
            author={author}
            key={author.id}
            index={index}
            deleteAuthor={deleteAuthor}
        />
    ));
    if (authorRows.length > 0) {
        return (
            <table className="table">
                <thead>
                <tr>
                    <th>&nbsp;</th>
                    <th>Name</th>
                </tr>
                </thead>
                <tbody>{authorRows}</tbody>
            </table>
        );
    } else if (!loading) {
        return <EmptyPage text="No Authors to display. Please add an author to view."/>;
    } else {
        return <h3>Loading <LoadingDots interval={100} dots={20} /></h3>;
    }
};

AuthorList.propTypes = {
    authors: PropTypes.array.isRequired,
    loading: PropTypes.bool.isRequired,
    deleteAuthor: PropTypes.func.isRequired
};
export default AuthorList;
