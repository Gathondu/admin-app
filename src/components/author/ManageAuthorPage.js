import React, {Fragment, PureComponent} from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { Prompt } from "react-router-dom";
import * as actions from "../../actions/authors";
import toastr from "toastr";
import TextInput from "../common/TextInput";

export class ManageAuthorPage extends PureComponent {
  state = {
    author: Object.assign({}, this.props.author),
    loading: false,
    errors: {firstName: "", lastName: ""},
    formIsValid: false,
    firstNameValid: false,
    lastNameValid: false,
    formUpdated: false
  };

  validateInput = author => {
    let errors = {};
    for (var field in author) {
      if (author.hasOwnProperty(field)) {
        switch (field) {
          case "firstName":
            if (author["firstName"].length > 0) {
              this.setState({firstNameValid: true});
            } else {
              errors["firstName"] =
                  "First Name should contain more than 1 character";
            }
            break;
          case "lastName":
            if (author["lastName"].length > 0) {
              this.setState({lastNameValid: true});
            } else {
              errors["lastName"] =
                  "Last Name should contain more than 1 character";
            }
            break;
          default:
            break;
        }
      }
    }
    if (Object.keys(errors).length > 0) {
      this.setState({errors: errors});
    } else {
      this.setState({errors: {}, formIsValid: true});
    }
  };

  updateAuthorState = e => {
    const field = e.target.name;
    let author = Object.assign({}, this.state.author);
    author[field] = e.target.value;
    this.validateInput(author);
    this.setState({author: author});
  };

  saveAuthor = e => {
    e.preventDefault();
    this.setState({loading: true});
    this.validateInput(this.state.author);
    if (this.state.formIsValid) {
      this.setState({formUpdated: false});
      this.props.actions
          .saveAuthor(this.state.author)
          .then(() => this.redirect())
          .catch(error => {
            this.setState({loading: false});
            toastr.error(error);
          });
    }
  };

  redirect = () => {
    this.setState({loading: false});
    toastr.success("Author Saved.");
    this.props.history.push("/authors");
  };

  render() {
    return <Fragment>
      <Prompt
          key="prompt"
          when={this.state.formUpdated}
          message="The form isn't saved, are you sure you want to leave the page?"
      />
      <form>
        <h1>Manage Author</h1>
        <TextInput
            label="First Name"
            onChange={this.updateAuthorState}
            name="firstName"
            value={this.state.author.firstName}
            error={this.state.errors.firstName}
            placeholder="Enter First Name"/>
        <TextInput
            label="Last Name"
            onChange={this.updateAuthorState}
            name="lastName"
            value={this.state.author.lastName}
            error={this.state.errors.lastName}
            placeholder="Enter Course Category"/>
        <input
            type="submit"
            disabled={this.state.loading || !this.state.formIsValid}
            value={this.state.loading ? 'Saving...' : 'Save'}
            className="btn btn-primary"
            onClick={this.saveAuthor}/>
      </form>
    </Fragment>
  }
}

export const getAuthorById = (authors, id) => {
    const author = authors.filter(author => author.id === id);
    if (author.length) return author[0];
    return null;
};
export const mapStateToProps = (state, ownProps) => {
    const authorId = ownProps.match.params.id;
    let author = {
        firstName: "",
        lastName: ""
    };
    if (authorId && state.authors.length > 0) {
        author = getAuthorById(state.authors, authorId);
    }
    return {
        author: author
    };
};
export const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(actions, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ManageAuthorPage);
