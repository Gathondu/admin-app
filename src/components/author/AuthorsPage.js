import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import AuthorList from "./AuthorList";
import  {deleteAuthor} from "../../actions/authors";
import {connect} from "react-redux";

export class AuthorsPage extends Component {
    handleDeleteAuthor = (id) => {
        this.props.deleteAuthor(id)
    };
    render () {
        const {authors, loading, history} = this.props;
        return <Fragment>
            <h1 key="title">Authors</h1>
            <input
                key="submit"
                type="submit"
                value="Add Author"
                className="btn btn-primary"
                onClick={() => history.push('/author')}
            />
            <AuthorList
                key="author-list"
                authors={authors}
                deleteAuthor={this.handleDeleteAuthor}
                loading={loading}
            />
        </Fragment>
    }
}

AuthorsPage.propTypes = {
    authors: PropTypes.array.isRequired,
    deleteAuthor: PropTypes.func.isRequired,
    loading: PropTypes.bool,
    history: PropTypes.shape({
        push: PropTypes.func.isRequired
    }).isRequired
};
const mapStateToProps = ({authors, totalAuthors, ajaxCallsInProgress}) => (
    {
        authors,
        loading: ajaxCallsInProgress > 0,
        totalAuthors
    }
);

export default connect(mapStateToProps, {deleteAuthor})(AuthorsPage);
