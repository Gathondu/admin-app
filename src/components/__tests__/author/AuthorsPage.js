import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {AuthorsPage} from '../../author/AuthorsPage';


describe('Manage Author Page component', () => {
    const setup = () => {
        const props = {
            authors: [],
            loading: false,
            deleteAuthor: jest.fn(),
            history: {
                push: jest.fn()
            }
        };

        return {
            props,
            presentationWrapper: shallow(<AuthorsPage { ...props } />)
        };
    };
    it('should render correctly', () => {
        const {presentationWrapper} = setup();
        expect(toJson(presentationWrapper)).toMatchSnapshot();
    });
    it('should redirect to /author page', () => {
        const {presentationWrapper} = setup();
        const handleDeleteSpy = jest.spyOn(
            presentationWrapper.instance(), 'handleDeleteAuthor'
        );
        presentationWrapper.instance().handleDeleteAuthor('cory-house');
        expect(handleDeleteSpy).toHaveBeenCalled()
    });

    it('should redirect to /author page', () => {
        const {presentationWrapper, props} = setup();
        presentationWrapper.find('input').simulate('click');
        expect(props.history.push).toHaveBeenCalled()
    });
});