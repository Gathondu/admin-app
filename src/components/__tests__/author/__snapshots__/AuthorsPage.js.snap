// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manage Author Page component should render correctly 1`] = `
<Fragment>
  <h1
    key="title"
  >
    Authors
  </h1>
  <input
    className="btn btn-primary"
    key="submit"
    onClick={[Function]}
    type="submit"
    value="Add Author"
  />
  <Connect(AuthorList)
    key="author-list"
  />
</Fragment>
`;
